src/Elliptic/Systems/Xcts/Equations.cpp

```diff
@@ -0,0 +1,619 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Elliptic/Systems/Xcts/Equations.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <functional>
+#include <optional>
+#include <string>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/TempBuffer.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Elliptic/Systems/Xcts/Geometry.hpp"
+#include "PointwiseFunctions/GeneralRelativity/IndexManipulation.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Literals.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace Xcts {
+namespace detail {
+void fully_contract(const gsl::not_null<Scalar<DataVector>*> result,
+                    const tnsr::II<DataVector, 3>& tensor1,
+                    const tnsr::II<DataVector, 3>& tensor2) noexcept {
+  get(*result) = tensor1.multiplicity(0_st) * tensor1[0] * tensor2[0];
+  for (size_t i = 1; i < tensor1.size(); ++i) {
+    get(*result) += tensor1.multiplicity(i) * tensor1[i] * tensor2[i];
+  }
+}
+void fully_contract(const gsl::not_null<Scalar<DataVector>*> result,
+                    const tnsr::II<DataVector, 3>& tensor1,
+                    const tnsr::II<DataVector, 3>& tensor2,
+                    const tnsr::ii<DataVector, 3>& metric) noexcept {
+  get(*result) = 0.;
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j < 3; ++j) {
+      for (size_t k = 0; k < 3; ++k) {
+        for (size_t l = 0; l < 3; ++l) {
+          get(*result) += metric.get(i, k) * metric.get(j, l) *
+                          tensor1.get(i, j) * tensor2.get(k, l);
+        }
+      }
+    }
+  }
+}
+}  // namespace detail
+
+void add_hamiltonian_sources(
+    const gsl::not_null<Scalar<DataVector>*> hamiltonian_constraint,
+    const Scalar<DataVector>& energy_density,
+    const Scalar<DataVector>& extrinsic_curvature_trace,
+    const Scalar<DataVector>& conformal_factor) noexcept {
+  get(*hamiltonian_constraint) +=
+      (square(get(extrinsic_curvature_trace)) / 12. -
+       2. * M_PI * get(energy_density)) *
+      pow<5>(get(conformal_factor));
+}
+
+void add_linearized_hamiltonian_sources(
+    const gsl::not_null<Scalar<DataVector>*> linearized_hamiltonian_constraint,
+    const Scalar<DataVector>& energy_density,
+    const Scalar<DataVector>& extrinsic_curvature_trace,
+    const Scalar<DataVector>& conformal_factor,
+    const Scalar<DataVector>& conformal_factor_correction) noexcept {
+  get(*linearized_hamiltonian_constraint) +=
+      ((5. / 12.) * square(get(extrinsic_curvature_trace)) -
+       10. * M_PI * get(energy_density)) *
+      pow<4>(get(conformal_factor)) * get(conformal_factor_correction);
+}
```

Comment:
Opened https://github.com/sxs-collaboration/spectre/issues/3071 for this. Pre-computing these quantities could potentially speed up current solves because they explicitly build the subdomain operator by evaluating the linearization many times, but the performance gain might decrease with better preconditioners.